src/openrct2-ui/scripting/CustomMenu.cpp

```diff
@@ -33,8 +46,169 @@ namespace OpenRCT2::Scripting
 
     void InitialiseCustomMenuItems(ScriptEngine& scriptEngine)
     {
-        scriptEngine.SubscribeToPluginStoppedEvent([](std::shared_ptr<Plugin> plugin) -> void { RemoveMenuItems(plugin); });
+        scriptEngine.SubscribeToPluginStoppedEvent(
+            [](std::shared_ptr<Plugin> plugin) -> void { RemoveMenuItemsAndTool(plugin); });
+    }
+
+    void CustomTool::OnUpdate(const ScreenCoordsXY& screenCoords)
+    {
+        InvokeEventHandler(onMove, screenCoords);
     }
+
+    void CustomTool::OnDown(const ScreenCoordsXY& screenCoords)
+    {
+        MouseDown = true;
+        InvokeEventHandler(onDown, screenCoords);
+    }
+
+    void CustomTool::OnDrag(const ScreenCoordsXY& screenCoords)
+    {
+    }
+
+    void CustomTool::Start()
+    {
+        auto& scriptEngine = GetContext()->GetScriptEngine();
+        scriptEngine.ExecutePluginCall(Owner, onStart, {}, false);
+    }
+
+    void CustomTool::OnUp(const ScreenCoordsXY& screenCoords)
+    {
+        MouseDown = false;
+        InvokeEventHandler(onUp, screenCoords);
+    }
+
+    void CustomTool::OnAbort()
+    {
+        auto& scriptEngine = GetContext()->GetScriptEngine();
+        scriptEngine.ExecutePluginCall(Owner, onFinish, {}, false);
+    }
+
+    static DukValue CoordsXYToDuk(duk_context* ctx, const CoordsXY& coords)
+    {
+        DukObject dukCoords(ctx);
+        dukCoords.Set("x", coords.x);
+        dukCoords.Set("y", coords.y);
+        return dukCoords.Take();
+    }
+
+    static DukValue CoordsXYToDuk(duk_context* ctx, const ScreenCoordsXY& coords)
```

Comment:
I'd either call these two `ToDuk` or specify that the second one is `ScreenCoordsXYToDuk`